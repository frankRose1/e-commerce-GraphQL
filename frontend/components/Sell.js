import React, { Component } from 'react';
import {Mutation} from 'react-apollo';
import Router from 'next/router';
import gql from 'graphql-tag';
import Form from './styles/Form';
import Error from './ErrorMessage';
import SubTitle from './styles/SubTitleStyles';


const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION($title: String!, $description: String!, $price: Int!, $image: String, $largeImage: String){
    createItem(title: $title, description: $description, price: $price, image: $image, largeImage: $largeImage){
      id
    }
  }
`;

class Sell extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0
  }

  uploadFile = async e => {
    const files = e.target.files;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'backpacks');

    const res = await fetch('https://api.cloudinary.com/v1_1/dbviuyhby/image/upload', {
      method: 'POST',
      body: data
    });
    const file = await res.json();
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url
    });
  }

  handleInputChange = e => {
    const {name, value, type} = e.target;
    const val = type === 'number' ? parseFloat(value) : value;
    this.setState({ [name]: val });
  }

  render() {
    return (
      <Mutation 
        mutation={CREATE_ITEM_MUTATION}
        variables={this.state}>
        {(createItem, {error, loading}) => (
          <>
            <SubTitle>List an Item For Sale</SubTitle>
            <Form
              data-test="sell-form"
              method="post" 
              onSubmit={async e => {
              e.preventDefault();

              const res = await createItem();
              this.setState({
                title: '',
                description: '',
                image: '',
                largeImage: '',
                price: 0
              });
              
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id }
              });
            }}>
            
              <fieldset disabled={loading} aria-busy={loading}>
              <Error error={error}/>
              <label htmlFor="file">
                  Image
                  <input 
                    type="file" 
                    id="file"
                    name="file"
                    placeholder="Add an image"
                    onChange={this.uploadFile}
                    required/>
                    {this.state.image && <img width="200" src={this.state.image} al="Upload Preview" />}
                </label>

                <label htmlFor="title">
                  Title
                  <input 
                    type="text" 
                    id="title"
                    name="title"
                    placeholder="Title"
                    value={this.state.title}
                    onChange={this.handleInputChange}
                    required/>
                </label>

                <label htmlFor="price">
                  Price
                  <input 
                    type="number" 
                    id="price"
                    name="price"
                    placeholder="Price"
                    value={this.state.price}
                    onChange={this.handleInputChange}
                    required/>
                </label>

                <label htmlFor="description">
                  Description
                  <textarea 
                    type="text" 
                    id="description"
                    name="description"
                    placeholder="Enter a description"
                    value={this.state.description}
                    onChange={this.handleInputChange}
                    required/>
                </label>

                <button type="submit">Creat{loading ? 'ing' : 'e'} Item!</button>

              </fieldset>
            </Form>
          </>
        )}
      </Mutation>
    );
  }
}

export default Sell;
export {CREATE_ITEM_MUTATION};