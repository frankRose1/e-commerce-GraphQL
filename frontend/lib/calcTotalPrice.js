/**
 * Take in a users cart and return the total cost of all items in cart
 * @param {array} cart - array of cartItems which contain data about a specific item in the DB
 * @return {integer} totalPrice - total of cart items
 */
function calcTotalPrice(cart){
  return cart.reduce((tally, cartItem) => {
    //in the case an item is deleted from the DB the cartItem will have an ID that points to a deleted item
    if (!cartItem.item) return tally;
    return tally + (cartItem.quantity * cartItem.item.price);
  }, 0);
}

export default calcTotalPrice;