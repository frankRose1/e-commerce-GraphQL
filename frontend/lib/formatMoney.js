/**
 * Format dollar amount of an item
 * @param {integer} amount - dollar amount of an item in cents
 * @return newly formatted price as a string, US currency 
 */
export default function(amount){
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  }
  //for whole dollar amounts, leave off the .00
  if (amount % 100 === 0) options.minimumFractionDigits = 0;
  const formatter = new Intl.NumberFormat('en-US', options);
  return formatter.format(amount / 100);
}