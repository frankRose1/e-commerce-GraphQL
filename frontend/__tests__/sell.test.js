import { mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import wait from 'waait';
import Sell, { CREATE_ITEM_MUTATION } from '../components/Sell';
import { fakeItem } from '../lib/testUtils';

Router.router = { push() {} };

const fakeImage = 'https://pizza.com/pizza.jpg';

//mock uploading images to cloudinary via fetch
global.fetch = jest.fn().mockResolvedValue({
  json: () => ({
    secure_url: fakeImage,
    eager: [{ secure_url: fakeImage }]
  })
});

describe('<Sell />', () => {
  it('renders and matches snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <Sell />
      </MockedProvider>
    );
    const sellForm = wrapper.find('form[data-test="sell-form"]');
    expect(toJSON(sellForm)).toMatchSnapshot();
  });

  it('uploads images on change event', async () => {
    const wrapper = mount(
      <MockedProvider>
        <Sell />
      </MockedProvider>
    );
    const input = wrapper.find('input[type="file"]');
    input.simulate('change', { target: { files: ['pizza.jpg'] } });
    await wait();
    //grab an instance of the component to be able to look at state
    const componentInstance = wrapper.find('Sell').instance();
    expect(componentInstance.state.image).toEqual(fakeImage);
    expect(componentInstance.state.largeImage).toEqual(fakeImage);
    expect(global.fetch).toHaveBeenCalled();
    global.fetch.mockReset();
  });

  it('updates state on change events', async () => {
    const wrapper = mount(
      <MockedProvider>
        <Sell />
      </MockedProvider>
    );
    wrapper
      .find('#title')
      .simulate('change', {
        target: { name: 'title', value: 'A new Item Title' }
      });
    wrapper
      .find('#price')
      .simulate('change', {
        target: { name: 'price', value: 5012, type: 'number' }
      });
    wrapper
      .find('#description')
      .simulate('change', {
        target: { name: 'description', value: 'a new item description!!!!' }
      });
    //see if state is equal to the changes in the inputs
    const componentInstance = wrapper.find('Sell').instance();
    expect(componentInstance.state).toMatchObject({
      title: 'A new Item Title',
      price: 5012,
      description: 'a new item description!!!!'
    });
  });

  it('creates an item on form submit', async () => {
    const item = fakeItem();
    const mocks = [
      {
        request: {
          query: CREATE_ITEM_MUTATION,
          variables: {
            title: item.title,
            price: item.price,
            description: item.description,
            image: '',
            largeImage: ''
          }
        },
        result: {
          data: { createItem: { ...item } }
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Sell />
      </MockedProvider>
    );
    //simulate someone filling out the form
    wrapper
      .find('#title')
      .simulate('change', { target: { name: 'title', value: item.title } });
    wrapper
      .find('#price')
      .simulate('change', {
        target: { name: 'price', value: item.price, type: 'number' }
      });
    wrapper
      .find('#description')
      .simulate('change', {
        target: { name: 'description', value: item.description }
      });
    //simulate the router pushing to the item page
    Router.router.push = jest.fn();
    wrapper.find('form').simulate('submit');
    await wait(50);
    expect(Router.router.push).toHaveBeenCalled();
    expect(Router.router.push).toHaveBeenCalledWith({
      pathname: '/item',
      query: { id: 'abc123' }
    });
  });
});
