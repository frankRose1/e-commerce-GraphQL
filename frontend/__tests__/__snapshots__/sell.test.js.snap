// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sell /> renders and matches snapshot 1`] = `
<form
  className="Form-sc-14226r0-0 iyeUlE"
  data-test="sell-form"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="file"
    >
      Image
      <input
        id="file"
        name="file"
        onChange={[Function]}
        placeholder="Add an image"
        required={true}
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="price"
    >
      Price
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="description"
    >
      Description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter a description"
        required={true}
        type="text"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Creat
      e
       Item!
    </button>
  </fieldset>
</form>
`;
